Use more C++ Concepts in WebCore

iidmsa opened the pull request:
#### 286bc2f2fea77c9b436308d367630adbceacf939
<pre>
Use more C++ Concepts in WebCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=293810">https://bugs.webkit.org/show_bug.cgi?id=293810</a>
<a href="https://rdar.apple.com/problem/152321345">rdar://problem/152321345</a>

Reviewed by Sam Weinig.

Replace std::enable_if with C++20 requires clauses for template constraints.
This should provide better readability and better compile-time performance by utilizing
more core language features.

* Source/WebCore/platform/graphics/Color.h:
* Source/WebCore/platform/graphics/ColorModels.h:
* Source/WebCore/platform/graphics/ImagePaintingOptions.h:
* Source/WebCore/platform/graphics/cairo/CairoOperationRecorder.cpp:
(WebCore::Cairo::createCommand):

Canonical link: <a href="https://commits.webkit.org/295661@main">https://commits.webkit.org/295661@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/03f3377d826e1070211a6e4e1ac9c19f76d37c29

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105789 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25540 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15933 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110986 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56385 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107830 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26118 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34043 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80337 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108795 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20489 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95463 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60650 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20169 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13560 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55824 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89906 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13599 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113835 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32929 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24266 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89416 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33293 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91694 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89086 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22712 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33962 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11768 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28410 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32854 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38265 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32600 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35949 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34198 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/platform/graphics/Color.h:
```diff
@@ -88,10 +88,12 @@ class Color {
 
-    template<typename ColorType, typename std::enable_if_t<IsColorTypeWithComponentType<ColorType, float>>* = nullptr>
+    template<typename ColorType>
+    requires IsColorTypeWithComponentType<ColorType, float>
```
Generally when using requires clauses I have been indenting them one further than prior line. In this case:

```c++
    template<typename ColorType>
        requires IsColorTypeWithComponentType<ColorType, float>
```

Though in this case, I think you can just make this by ensuring IsColorTypeWithComponentType is a proper concept:

```c++
    template<IsColorTypeWithComponentType<float> ColorType>
```

(see CSSPrimitiveNumericUnits.h for an example).

We should standardize and document it in https://www.webkit.org/code-style-guidelines/)

weinig replied:
(Generally, I think the non-requires syntax should be preferred when possible).

iidmsa replied:
Thank you for your comment.

weinig commented on Source/WebCore/platform/graphics/cairo/CairoOperationRecorder.cpp:
```diff
@@ -62,14 +62,18 @@ struct OperationData {
 
 template<typename T, typename... Args>
-auto createCommand(Args&&... arguments) -> std::enable_if_t<std::is_base_of<OperationData<std::decay_t<Args>...>, T>::value, std::unique_ptr<PaintingOperation>> {
+    requires std::is_base_of_v<OperationData<std::decay_t<Args>...>, T>
```
Does the standard std::derived_from concept work here?

iidmsa replied:
Yes, 

```
template<typename T, typename... Args>
requires std::derived_from<T, OperationData<std::decay_t<Args>...>>
```

weinig commented on Source/WebCore/platform/graphics/cairo/CairoOperationRecorder.cpp:
```diff
@@ -62,14 +62,18 @@ struct OperationData {
 
 template<typename T>
-auto createCommand() -> std::enable_if_t<std::is_base_of<OperationData<>, T>::value, std::unique_ptr<PaintingOperation>> {
+    requires std::is_base_of_v<OperationData<>, T>
```
Does the standard std::derived_from concept work here?

iidmsa replied:
Yes, 

```
template<typename T, typename... Args>
requires std::derived_from<T, OperationData<std::decay_t<Args>...>>
```

iidmsa committed:
Use more C++ Concepts in WebCore
https://bugs.webkit.org/show_bug.cgi?id=293810
rdar://problem/152321345

Reviewed by Sam Weinig.

Replace std::enable_if with C++20 requires clauses for template constraints.
This should provide better readability and better compile-time performance by utilizing
more core language features.

* Source/WebCore/platform/graphics/Color.h:
* Source/WebCore/platform/graphics/ColorModels.h:
* Source/WebCore/platform/graphics/ImagePaintingOptions.h:
* Source/WebCore/platform/graphics/cairo/CairoOperationRecorder.cpp:
(WebCore::Cairo::createCommand):

Canonical link: https://commits.webkit.org/295661@main
